qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py

```diff
@@ -0,0 +1,145 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import Dict, Set, TYPE_CHECKING
+
+import cirq
+import numpy as np
+import sympy
+from attrs import field, frozen
+from numpy.typing import NDArray
+
+from qualtran import bloq_example, Controlled, CtrlSpec, GateWithRegisters, Signature
+from qualtran.bloqs.basic_gates import SU2RotationGate
+from qualtran.bloqs.generalized_qsp import GeneralizedQSP
+from qualtran.bloqs.qsp.polynomial_approximations import (
+    approx_exp_cos_by_jacobi_anger,
+    degree_jacobi_anger_approximation,
+)
+from qualtran.bloqs.qubitization_walk_operator import QubitizationWalkOperator
+from qualtran.resource_counting.symbolic_counting_utils import SymbolicFloat, SymbolicInt
+
+if TYPE_CHECKING:
+    from qualtran import BloqBuilder, SoquetT
+    from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+@frozen
+class HamiltonianSimulationByGQSP(GateWithRegisters):
+    r"""Hamiltonian simulation using Generalized QSP given a qubitized quantum walk operator.
+
+    Implements Hamiltonian simulation given a walk operator from SELECT and PREPARE oracles.
+
+    We can use the Jacobi-Anger expansion to obtain low-degree polynomial approximations for the $\cos$ function:
+
+        $$ e^{it\cos\theta} = \sum_{n = -\infty}^{\infty} i^n J_n(t) (e^{i\theta})^n $$
+
+    where $J_n$ is the $n$-th [Bessel function of the first kind](https://en.wikipedia.org/wiki/Bessel_function#Bessel_functions_of_the_first_kind), which is provided by `scipy.special.jv`.
+    We can cutoff at $d = O(t + \log(1/\epsilon) / \log\log(1/\epsilon))$ to get an $\epsilon$-approximation (Theorem 7):
+
+        $$ P[t](z) = \sum_{n = -d}^d i^n J_n(t) z^n $$
+
+    References:
+        [Generalized Quantum Signal Processing](https://arxiv.org/abs/2308.01501)
+            Motlagh and Wiebe. (2023). Theorem 7.
+    """
+
+    walk_operator: QubitizationWalkOperator
+    t: SymbolicFloat = field(kw_only=True)
+    alpha: SymbolicFloat = field(kw_only=True)
+    precision: SymbolicFloat = field(kw_only=True)
+
+    def _parameterized_(self):
+        return cirq.is_parameterized((self.t, self.alpha, self.precision))
```

Comment:
You can also use `is_symbolic(self.t, self.alpha, self.precision)` where `is_symbolic` is defined in qualtran/resource_counting/symbolic_counting_utils.py; to reduce the dependence on cirq